wordpress-mobile/WordPress-iOS
# Stats: PeriodStore + improve CoreData Persistence + Caching

This PR adds functionality to fix a caching issue in the PeriodStore

The StatsPeriodStore has a [note](https://github.com/wordpress-mobile/WordPress-iOS/blob/246e149ece9c94c3a773739302ba4d7e452df389/WordPress/Classes/Stores/StatsPeriodStore.swift#L580) on how the persist to CoreData has an unintuitive call.  And after diving deeply + trying to diagnose I can see that at the tail end of fetchAsyncData no further saves to CoreData are called

### Attempted solution 1

~~I've tried to fix this by adding a barrierblock as the last operation to fetchAsyncData but at this time that is still not working so I've left this in draft status~~

~~Note that all the other fetch calls in StatsPeriodStore also save to core data ie: fetchAllVideos, fetchAllSearchTerms etc so I'm not sure why fetchAsyncData did not have this as well and cannot see why it shouldn't have (this call)~~

### Attempted solution 2

The barrierblock solution was discarded in favor of using dispatch groups to save to core data once multiple fetches have completed.

Fixes #19430

These should be the steps for a successful test (same as in the issue)
1. Enable new stats feature flags
2. Go to Stats screen and add Views & Visitors card
3. Tap back and return to main stats screen
4. The total numbers on Views & Visitors should hold their cached value

## Regression Notes
1. Potential unintended areas of impact
Stats

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing

3. What automated tests I added (or what prevented me from doing so)

PR submission checklist:
